neps/nep-0568.md

```diff
@@ -187,18 +197,39 @@ In this NEP, we propose updating the ShardId semantics to allow for arbitrary id
 
 ## Reference Implementation
 
-```text
-[This technical section is required for Protocol proposals but optional for other categories. A draft implementation should demonstrate a minimal implementation that assists in understanding or implementing this proposal. Explain the design in sufficient detail that:
+```
+should_split_shard(block):
+  shard_layout = epoch_manager.shard_layout(block.epoch_id())
+  next_shard_layout = epoch_manager.shard_layout(block.next_epoch_id())
+  return epoch_manager.is_next_block_epoch_start(block) && shard_layout != next_shard_layout
 
-* Its interaction with other features is clear.
-* Where possible, include a Minimum Viable Interface subsection expressing the required behavior and types in a target programming language. (ie. traits and structs for rust, interfaces and classes for javascript, function signatures and structs for c, etc.)
-* It is reasonably clear how the feature would be implemented.
-* Corner cases are dissected by example.
-* For protocol changes: A link to a draft PR on nearcore that shows how it can be integrated in the current code. It should at least solve the key technical challenges.
+on chain.postprocess_block(block):
+  next_shard_layout = epoch_manager.shard_layout(block.next_epoch_id())
+  if should_split_shard(block):
+    resharding_manager.split_shard(split_shard_event, next_shard_layout)
 
-The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.]
+on resharding_manager.split_shard(split_shard_event, next_shard_layout):
+  set State mapping
+  start FlatState resharding
+  process MemTrie resharding:
+    freeze MemTrie, create HybridMemTries
+    for each child shard:
+      mem_tries[shard].retain_split_shard(boundary_account)
+
+mem_trie.retain_split_shard(boundary_account):
+  split shard by path as described above while generating the proof
+  saving the proof as state transition for pair (block, new_shard_uid)
+
+then, the proof is sent as one of implicit transitions in ChunkStateWitness
+
+then, on chunk validation path, chunk validator understands if resharding is a part of state transition, using the same should_split_shard condition
+
+and then it calls Trie(state_transition_proof).retain_split_shard(boundary_account) which should succeed if proof is sufficient and generates new state root
+
+finally, it checks that the new state root matches the state root proposed in ChunkStateWitness. if the whole ChunkStateWitness is valid, then chunk validator sends endorsement which also endorses the resharding.
```

Comment:
We may want to take out these long text lines from the code block, or newline them manually